### ISSUE TITLE
Figure out how to sign remote packages

### ISSUE THREAD

**Issue by phenaproxima** [php-tuf/php-tuf#128]
#127 introduces a small under-the-hood change to make it possible for php-tuf/composer-integration to use TUF to download remote targets. A "remote target" is any target which is NOT hosted on the same server as the TUF metadata. Essentially it's a concept of treating an arbitrary URL as a signed, validated target.

As far as I can tell, we're going to _need_ to support something like this in order for us to sign packages which come from Packagist, rather than drupal.org. But it doesn't look like the Python repository tool (which we use to create TUF metadata from arbitrary, local target files) has a way to sign a remote URL. It seems to expect everything it signs to be a local file. Which makes sense, but doesn't help us.

The workaround I'm implementing to sign remote targets is:

1. Download the remote file from some URL that you trust.
2. Rename the downloaded file after the SHA256 hash of the URL from which you retrieved it. So basically, `somefile.zip` --> `sha256HashOfTheUrl`.
3. Modify generate.py to sign that target.

Then, the plugin is doing this:

1. Use a custom package loader to store the SHA256 hash of the package's dist URL -- i.e., determine the name of the target.
2. Use a custom file fetcher, extending the Guzzle fetcher, which stores a map of target names to "real" URLs.
3. At download time, ensure that the real URL is downloaded and run through TUF's validation.

In other words: TUF knows about the remote target by a SHA256 hash. But internally, the fetcher knows to retrieve a different URL. Ugly, I know...but until the Python repository tool can sign remote URLs natively, this is necessary. I would massively prefer if TUF could just treat URLs like regular targets!

**Comment by davidstrauss** [comment]
> As far as I can tell, we're going to need to support something like this in order for us to sign packages which come from Packagist, rather than drupal.org.

I was also considering something like configuring Drupal.org Fastly to proxy.

**Comment by phenaproxima** [comment]
I think this is pretty much resolved at this point.

This library (and, as best I can figure, TUF itself) doesn't care where the data actually originates. As long as it matches the verified hashes available in the TUF metadata, it's trustworthy.

Signing remote packages is also not something this library will need to worry about. This merely verifies signed data, but doesn't sign anything itself. The Composer plugin uses the scheme of identifying a signed package by the target name with the scheme `VENDOR/NAME/NORMALIZED_VERSION`.

I'm closing this out.
